### ISSUE TITLE
PERF: Solution.WithDocumentText takes ~20% of UI thread time under TextBufferEdit.Apply in razor files

### ISSUE THREAD

**Issue by ToddGrun** [dotnet/roslyn#45430]
During analysis of the razor perf test, I noticed that Solution.WithDocumentText is taking more than a fair portion of time during text buffer edits. Most of the time in WithDocumentText is spent in CompilationTracker+State.GetUnrootedSymbols adding to a ConditionalWeakTable which ends up in clr!CrstBase::SpinEnter.

The question here is whether either the expensive work done during Solution.WithDocumentText can be moved off the UI thread or modified to not be so expensive.

Razor perf test: https://devdivperf/TestTrend.aspx#GT=Test&Sku=70&BID=267439&SKU2=70&BID2=267438&CID=3125775&HE=1805&SE=23&HE2=1805&SE2=23&OS=128&OS2=128&LM=677&LM2=677&SO=237&SO2=237&SID=68846&TcName=WebTools.TypingInRazorView&SH=0

**Comment by jasonmalinowski** [comment]
Tagging @CyrusNajmabadi who can speak more to what GetUnrootedSymbols is doing here -- it looks like we're using a CWT as a list of weak references, and if the CWT is the source of overhead we might be able to do something cheaper there. Or maybe it can be deferred more.

**Comment by jasonmalinowski** [comment]
And the traces that @ToddGrun had implied that the work being done in WithDocumentText is when we're forking a new Solution because of a text change in the buffer; we can't really not be on that path, but 80% of that time doing the forking was in GetUnrootedSymbols which seems excessive.

**Comment by CyrusNajmabadi** [comment]
Can you post the call stack that leads from you to teh call to GetUnrootedSymbols?

**Comment by CyrusNajmabadi** [comment]
@stephentoub Is there a good person to talk to to ask questions about ConditionalWeakTable?  Thanks!

**Comment by ToddGrun** [comment]
Name | Inc % | Inc
-- | -- | --
microsoft.codeanalysis.workspaces!Microsoft.CodeAnalysis.SolutionState+CompilationTracker+State.GetUnrootedSymbols(class Microsoft.CodeAnalysis.Compilation) | 94.8 | 1,175
+ microsoft.codeanalysis.workspaces.ni!Microsoft.CodeAnalysis.SolutionState+CompilationTracker+InProgressState..ctor(Microsoft.CodeAnalysis.Compilation, TrackedGeneratorDriver, System.Collections.Immutable.ImmutableArray`1>) | 94.7 | 1,174
\|+ microsoft.codeanalysis.workspaces.ni!Microsoft.CodeAnalysis.SolutionState+CompilationTracker+State.Create(Microsoft.CodeAnalysis.Compilation, TrackedGeneratorDriver, System.Collections.Immutable.ImmutableArray`1>) | 94.7 | 1,174
\| + microsoft.codeanalysis.workspaces.ni!Microsoft.CodeAnalysis.SolutionState+CompilationTracker.Fork(Microsoft.CodeAnalysis.ProjectState, CompilationAndGeneratorDriverTranslationAction, Boolean, System.Threading.CancellationToken) | 84.0 | 1,041
\| \|+ microsoft.codeanalysis.workspaces.ni!SolutionState.ForkProject | 84.0 | 1,041
\| \| + microsoft.codeanalysis.workspaces.ni!SolutionState.UpdateDocumentState | 84.0 | 1,041
\| \|  + microsoft.codeanalysis.workspaces.ni!SolutionState.WithDocumentText | 84.0 | 1,041
\| \|   + microsoft.codeanalysis.workspaces.ni!Solution.WithDocumentText | 84.0 | 1,041
\| \|    + microsoft.codeanalysis.workspaces!Microsoft.CodeAnalysis.Workspace+<>c.<OnDocumentTextChanged>b__85_1(class Microsoft.CodeAnalysis.Solution,class Microsoft.CodeAnalysis.DocumentId,class Microsoft.CodeAnalysis.Text.SourceText,value class Microsoft.CodeAnalysis.PreservationMode) | 83.9 | 1,040
\| \|    \|+ microsoft.codeanalysis.workspaces.ni!Workspace.OnDocumentTextChanged | 83.9 | 1,040
\| \|    \| + microsoft.codeanalysis.workspaces!Microsoft.CodeAnalysis.Workspace+<>c.<OnDocumentOpened>b__188_0(class Microsoft.CodeAnalysis.Workspace,class Microsoft.CodeAnalysis.DocumentId,class Microsoft.CodeAnalysis.Text.SourceText,value class Microsoft.CodeAnalysis.PreservationMode) | 83.9 | 1,040
\| \|    \|  + microsoft.codeanalysis.editorfeatures.text.ni!? | 83.9 | 1,040
\| \|    \|   + microsoft.visualstudio.platform.vseditor.ni!Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent[System.__Canon](System.Object, System.EventHandler`1, System.__Canon) | 83.9 | 1,040
\| \|    \|    + microsoft.visualstudio.platform.vseditor.ni!BaseBuffer.RawRaiseEvent | 83.9 | 1,040
\| \|    \|     + microsoft.visualstudio.platform.vseditor.ni!Microsoft.VisualStudio.Text.Implementation.BaseBuffer+TextContentChangedEventRaiser.RaiseEvent(Microsoft.VisualStudio.Text.Implementation.BaseBuffer, Boolean) | 83.9 | 1,040
\| \|    \|      + microsoft.visualstudio.platform.vseditor.ni!BufferGroup.FinishEdit | 83.9 | 1,040
\| \|    \|       + microsoft.visualstudio.platform.vseditor.ni!Microsoft.VisualStudio.Text.Implementation.BaseBuffer+TextBufferEdit.Apply() | 83.9 | 1,040

**Comment by ToddGrun** [comment]
that formatted oddly :)

**Comment by CyrusNajmabadi** [comment]
Thanks, super helpful.

**Comment by jasonmalinowski** [comment]
Minimally I wasn't sure if a simple List<WeakReference> would be cheaper here...

**Comment by CyrusNajmabadi** [comment]
> Or maybe it can be deferred more.

We could potentially defer more, but we'd have to be super super careful.  i.e. before we ever handed out a COmpilation, we would have to make sure that we initialized these values.

**Comment by stephentoub** [comment]
> Is there a good person to talk to to ask questions about ConditionalWeakTable? Thanks!

Depends on the question, but you can start with me; for anything deeper it's probably @Maoni0.

**Comment by CyrusNajmabadi** [comment]
> Minimally I wasn't sure if a simple List<WeakReference> would be cheaper here...

I imagine that would be fine.  It would mean some checks become linear isntead of O(1)... but likely wouldn't matter in practice.

**Comment by CyrusNajmabadi** [comment]
@stephentoub Do you have any insight on:

> Most of the time in WithDocumentText is spent in CompilationTracker+State.GetUnrootedSymbols adding to a ConditionalWeakTable which ends up in clr!CrstBase::SpinEnter.

@ToddGrun in your project, how many metadata references does your project have?  That number will be the number of items we're adding to the CWT.

--

Also, we use the CWT like this:

```c#
                public static ConditionalWeakTable<ISymbol, object?> GetUnrootedSymbols(Compilation compilation)
                {
                    var result = new ConditionalWeakTable<ISymbol, object?>();

                    var compAssembly = compilation.Assembly;
                    result.Add(compAssembly, null);

                    // The dynamic type is also unrooted (i.e. doesn't point back at the compilation or source
                    // assembly).  So we have to keep track of it so we can get back from it to a project in case the 
                    // underlying compilation is GC'ed.
                    if (compilation.Language == LanguageNames.CSharp)
                        result.Add(compilation.DynamicType, null);

                    foreach (var reference in compilation.References)
                    {
                        var symbol = compilation.GetAssemblyOrModuleSymbol(reference);
                        if (symbol == null)
                            continue;

                        result.Add(symbol, null);
                    }

                    return result;
                }
            }
```

basically we're just adding to it (and not sharing it), until the end when we pass it out.  Is there a better way to set that up that could sidestep the perf issue?

**Comment by sharwell** [comment]
> basically we're just adding to it (and not sharing it), until the end when we pass it out

I would not expect a problem with the CWT itself in this case. If the trace conclusively shows the time in CWT.Add, I would first look at the number of times `GetUnrootedSymbols` was called (or perhaps look at the number of times a CWT was created in the allocation traces).

**Comment by jasonmalinowski** [comment]
Also, do we have to compute that list eagerly or can we defer it a bit?

**Comment by ToddGrun** [comment]
Here are the perf test sources: https://devdiv.visualstudio.com/DevDiv/_git/VS?path=%2Fsrc%2FTests%2FDDRIT.Perf%2FWebTools%2FWebTools.cs&version=GBmaster

Here is the path to the sln opened as part of the test: https://devdiv.visualstudio.com/Pkgs/_git/VS.TestAssets.MusicStore?path=%2FAssets%2FMusicStore&version=GBmaster&_a=contents

**Comment by CyrusNajmabadi** [comment]
> Also, do we have to compute that list eagerly or can we defer it a bit?

Mentioned here: https://github.com/dotnet/roslyn/issues/45430#issuecomment-649078715

It can be defferred.  But it must happen prior to anyone else being able to witness any symbols from this compilation.

THis is tricky, esp. in P2P scenarios.  i.e. if we have project A and B (where B references A).  THen the Compilation for B will point at the Compilation for A.  If someone gets the Compilation for B, they can observe symbols from A, and we'd have to make sure this weak table was created before that happened.

**Comment by ToddGrun** [comment]
> Here are the perf test sources: https://devdiv.visualstudio.com/DevDiv/_git/VS?path=%2Fsrc%2FTests%2FDDRIT.Perf%2FWebTools%2FWebTools.cs&version=GBmaster
> 
> Here is the path to the sln opened as part of the test: https://devdiv.visualstudio.com/Pkgs/_git/VS.TestAssets.MusicStore?path=%2FAssets%2FMusicStore&version=GBmaster&_a=contents

Actually, here is the path to the project: https://devdiv.visualstudio.com/Pkgs/_git/VS.TestAssets.MusicStore?path=%2FAssets%2FMusicStore%2Fsamples%2FMusicStore

Also, the profiles are available from the link posted earlier: https://devdivperf/TestTrend.aspx#GT=Test&Sku=70&BID=267439&SKU2=70&BID2=267438&CID=3125775&HE=1805&SE=23&HE2=1805&SE2=23&OS=128&OS2=128&LM=677&LM2=677&SO=237&SO2=237&SID=68846&TcName=WebTools.TypingInRazorView&SH=0

**Comment by sharwell** [comment]
Note that it's easy to avoid `ConditionalWeakTable<TKey, TValue>` for this scenario. We just need to use `HashSet<ReferenceHolder>`, where `ReferenceHolder` is a struct that implements `IEquatable<ReferenceHolder>` and wraps either a `T` or a `WeakReference<T>`.

**Comment by stephentoub** [comment]
> adding to a ConditionalWeakTable which ends up in clr!CrstBase::SpinEnter

What is the full stack down to the SpinEnter?

**Comment by ToddGrun** [comment]
> > adding to a ConditionalWeakTable which ends up in clr!CrstBase::SpinEnter
> 
> What is the full stack down to the SpinEnter?


Name | Inc % | Inc
-- | -- | --
microsoft.codeanalysis.workspaces!Microsoft.CodeAnalysis.SolutionState+CompilationTracker+State.GetUnrootedSymbols(class Microsoft.CodeAnalysis.Compilation) | 2.2 | 999
+ mscorlib.ni!System.Runtime.CompilerServices.ConditionalWeakTable`2[System.__Canon,System.__Canon].Add(System.__Canon, System.__Canon) | 2.0 | 925
+ mscorlib.ni!System.Runtime.CompilerServices.ConditionalWeakTable`2[System.__Canon,System.__Canon].CreateEntry(System.__Canon, System.__Canon) | 2.0 | 901
\|+ clr!DependentHandle::nInitialize | 1.8 | 823
\| + clr!BaseDomain::CreateDependentHandle | 1.8 | 814
\| \|+ clr!GCHandleStore::CreateDependentHandle | 1.7 | 789
\| \| + clr!??TableAllocSingleHandleFromCache | 1.7 | 777
\| \|  + clr!TableCacheMissOnAlloc | 1.7 | 775
\| \|   + clr!CrstBase::SpinEnter | 1.7 | 757

**Comment by stephentoub** [comment]
> clr!DependentHandle::nInitialize

Ok, the contention here is inside the GC's handle store for dependency handles.  So a question for @Maoni0.

**Comment by sharwell** [comment]
I agree with https://github.com/dotnet/roslyn/issues/45430#issuecomment-649091584. With that said, I submitted a pull request to eliminate this table (and the use of dependent handles) altogther.
